dotenv-linter/src/checks/schema_validator.rs

```diff
@@ -0,0 +1,118 @@
+use std::collections::HashSet;
+
+use super::Check;
+use crate::{
+    common::{LintKind, Warning},
+    schema::SchemaValueType,
+};
+use dotenv_lookup::LineEntry;
+use email_address::EmailAddress;
+use url::Url;
+pub(crate) struct SchemaChecker<'a> {
+    schema: Option<&'a crate::schema::DotEnvSchema>,
+    seen_keys: HashSet<String>,
+    last_line_number: usize,
+}
+
+impl<'a> SchemaChecker<'a> {
+    pub fn new(opts: &'a crate::cli::options::CliOptions) -> Self {
+        Self {
+            schema: opts.schema.as_ref(),
+            seen_keys: HashSet::new(),
+            last_line_number: 0,
+        }
+    }
+}
+
+impl Check for SchemaChecker<'_> {
+    fn run(&mut self, line: &LineEntry) -> Option<Warning> {
+        self.schema?;
+        self.last_line_number = line.number;
+        let schema = self.schema.unwrap();
+        let key = line.get_key()?;
+        self.seen_keys.insert(key.to_string());
+
+        if let Some(entry) = schema.entries.iter().find(|i| i.key == key) {
```

Comment:
Probably would be nice to represent schema elements as a hashmap?

Just to optimize the access, instead of filtering the iterator. 
Just because we do this for each line of file actually 🤔

Reply:
I wondered about this, but could not work out how to tell serde to deseriliaze into a hashmap. I will have another pass at it.

Reply:
I think it would be correct to deserialize to current `Vec` based structure and then just map all elements to hashmap, probably using enums with sub attributes (e.g validation rules for string, number and etc)

Another alternative: instead of `array`, try to use `Object` in `JSON`, where `key` is env variable name 🤔

Reply:
Oh, I see, you already updated that!
Thanks

Left a suggestion for new code!